### PR TITLE
Support conformal mesh for PEC structures

### PR DESCRIPTION
**Pull request by weiliangjin2021**
Mode solver: the formulation of mode solver will follow the treatment in Benkler's paper (10.1109/TAP.2006.875909), where the split curl is treated as scaled permittivity and E-field. In the end of the mode solver, the E-field is scaled back. 

We also enable `incidence_matrix` when PEC subpixel is applied.

- [x] Support mu_cross
- [x] Support split_curl coefficient
- [x] Set fields to be 0 inside PEC

### CODE REVIEW

**Comment by weiliangjin2021** [comment]
Now time-stepping in the backend is working, and mode solver will be the next in my list. So it might be a good time to settle down on frontend designs.

- Currently, i add a field [`conformal_mesh_spec`](https://github.com/flexcompute/tidy3d/blob/2a4ef490e257ef88f6c6c1dbd05befd7751378c2/tidy3d/components/simulation.py#L192) in the simulation class, which only takes effect if subpixel=True. This field controls what type of conformal mesh scheme and specs to be applied to PEC interfaces. 
- This field takes [ConformalMeshSpec](https://github.com/flexcompute/tidy3d/blob/2a4ef490e257ef88f6c6c1dbd05befd7751378c2/tidy3d/components/grid/grid_spec.py#L25-L42) object, which specify the conformal mesh scheme, and the Courant number reduction rate.
- In this draft, there are 3 supported schemes: staricasing, YM, and Benkler. As discussed with @momchil-flex earlier, a better design will be: staircasing when subpixel=False, and conformal mesh scheme only takes 2 types: Benkler (or naming it as DM since it's more popular?), and heuristic approach in the old tidy3d. YM can be dropped as it doesn't seem to outperform staircasing in many cases.
- As for courant number reduction, it is only used in Benkler's scheme. The time step size is reduced with [this factor](https://github.com/flexcompute/tidy3d/blob/2a4ef490e257ef88f6c6c1dbd05befd7751378c2/tidy3d/components/simulation.py#L2167-L2172) (a [reduced courant factor](https://github.com/flexcompute/tidy3d/blob/2a4ef490e257ef88f6c6c1dbd05befd7751378c2/tidy3d/components/simulation.py#L2188) is used). In a regular Yee-cell, the upper bound of the courant number is scaled from sim.courant to sim.courant*(1-reduction). In a conformal cell, the upper bound is still sim.courant. 

Looking forward to your feedbacks @tylerflex @momchil-flex !

**Comment by tylerflex** [comment]
Hi @weiliangjin2021, this is exciting! I think it's a good idea to put some extra thought into this front end design in case we want to extend the conformal mesh to other mediums in the future? or add other schemes. I like what you've started with but have a few initial comments and suggestions.

In summary:
* Maybe we can simplify by removing staircasing and YM, which just leaves Benkler?
* If no conformal mesh, just leave the spec as None? and validate that if the spec is not nOne, subpixel is also on?
* We could use classes to differentiate the scheme types, with just one Benkler class for now, instead of using `scheme` type. Seems unnecessary because all of the logic seems to just do different things if `scheme == "Benkler"`.

> * Currently, i add a field [`conformal_mesh_spec`](https://github.com/flexcompute/tidy3d/blob/2a4ef490e257ef88f6c6c1dbd05befd7751378c2/tidy3d/components/simulation.py#L192) in the simulation class, which only takes effect if subpixel=True. This field controls what type of conformal mesh scheme and specs to be applied to PEC interfaces.
> * This field takes [ConformalMeshSpec](https://github.com/flexcompute/tidy3d/blob/2a4ef490e257ef88f6c6c1dbd05befd7751378c2/tidy3d/components/grid/grid_spec.py#L25-L42) object, which specify the conformal mesh scheme, and the Courant number reduction rate.

I think this sounds like a good approach overall.

Do we plan on eventually supporting other mediums besides PEC? and if so, will they need different specs? if both answers are yes, we could consider this field being called `pec_conformal_mesh_spec`? or similar?

> * In this draft, there are 3 supported schemes: staricasing, YM, and Benkler. As discussed with @momchil-flex earlier, a better design will be: staircasing when subpixel=False, and conformal mesh scheme only takes 2 types: Benkler (or naming it as DM since it's more popular?), and heuristic approach in the old tidy3d. YM can be dropped as it doesn't seem to outperform staircasing in many cases.

What is heuristic approach? would that just be what we do now? (non conformal?)

So basically we only will support 3 combinations of subpixel and mesh spec?

```
| subpixel | conformal mesh |
| -------- | -------------- |
| False    | NA             |
| True     | None  (?)      |
| True     | Benkler        |
```

maybe you can clarify for me, especially the 2nd row.

If so, I'd say maybe the way I'd approach this front end design wold be (vaguely):

```py
class ConformalMeshSpec(Tidy3DBaseModel, ABC):
     pass

class BenklerConformalMeshSpec(ConformalMeshSpec):
    courant_reduction : float = DEFAULT_COURANT_REDUCTION_BENKLER

ConformalMeshSpecType = Union[BenklerConformalMeshSpec]
```

in `Simulation` just validate that `subpixel` is True if `conformal_mesh_spec` is defined? 

And the courant to use as a property depending on the conformal mesh spec looks good to me. 

> * As for courant number reduction, it is only used in Benkler's scheme. The time step size is reduced with [this factor](https://github.com/flexcompute/tidy3d/blob/2a4ef490e257ef88f6c6c1dbd05befd7751378c2/tidy3d/components/simulation.py#L2167-L2172) (a [reduced courant factor](https://github.com/flexcompute/tidy3d/blob/2a4ef490e257ef88f6c6c1dbd05befd7751378c2/tidy3d/components/simulation.py#L2188) is used). In a regular Yee-cell, the upper bound of the courant number is scaled from sim.courant to sim.courant*(1-reduction). In a conformal cell, the upper bound is still sim.courant.

sounds good to me.

**Comment by weiliangjin2021** [comment]
> Do we plan on eventually supporting other mediums besides PEC? and if so, will they need different specs? if both answers are yes, we could consider this field being called pec_conformal_mesh_spec? or similar?

One plan is to support high conductivity medium, e.g. following [this approach](https://ieeexplore.ieee.org/document/5617234). That sounds good, since we can then have high_conductivity_conformal_mesh_spec to be not confused with pec_conformal_mesh_spec.

> Maybe we can simplify by removing staircasing and YM, which just leaves Benkler?

Now as I think about it more, I feel we should keep staircasing in conformal_mesh_spec. That would allow us to apply subpixel to dielectric medium, but staircasing to PEC. With that consideration, the table will be

| subpixel | conformal mesh | scheme applied |
| -------- | -------------- |-------------------|
| False    | any type            |Staircasing|
| True     | None        |Staircasing|
| True     | Heuristic      |Heuristic|
| True     | Benkler        |Benkler|

While what we have now (before this PR) is
| subpixel | scheme applied |
| -------- | -------------- |
| True     | Staircasing      |
| False     | Heuristic      |

**Comment by momchil-flex** [comment]
> Now as I think about it more, I feel we should keep staircasing in conformal_mesh_spec. That would allow us to apply subpixel to dielectric medium, but staircasing to PEC. With that consideration, the table will be
> 
> subpixel	conformal mesh	scheme applied
> False	any type	Staircasing
> True	None	Staircasing
> True	Staircasing	Staircasing
> True	Heuristic	Heuristic
> True	Benkler	Benkler

I think this makes sense, with the only comment that the second row doesn't seem needed then (i.e. `conformal_mesh=None` is not allowed, `Staircasing` is default).

File: tidy3d/plugins/mode/solver.py [hunk 1]
```diff
@@ -738,6 +781,38 @@ def eigs_to_effective_index(cls, eig_list: Numpy, mode_solver_type: ModeSolverTy
...
 
+    @staticmethod
+    def format_medium_data(mat_data):
+        """
+        mat_data can be either permittivity or permeability. It's either a single 2D array
+        defining the relative property in the cross-section, or nine 2D arrays defining
+        the property at the E(H)x, E(H)y, and E(H)z locations of the Yee grid in the order
+        xx, xy, xz, yx, yy, yz, zx, zy, zz.
+        """
+        if isinstance(mat_data, Numpy):
```
[the remark this answers is not in the document]
Comment (weiliangjin2021) [reply tidy3d/plugins/mode/solver.py:792]: the same applies to epsilon?

[the remark this answers is not in the document]
Comment (tylerflex) [reply tidy3d/plugins/mode/solver.py:792]: I thought epsilon was always 9 arrays? maybe not. Anyway, this is just a suggestion, in my experience when functions accept different input types and try to handle it, things can get messy. Maybe in this case it's ok so I guess use your best judgement.

[the remark this answers is not in the document]
Comment (tylerflex) [reply tidy3d/plugins/mode/solver.py:792]: ok, it's not a big deal, especially if it already was like that. feel free to ignore

File: tidy3d/components/grid/grid_spec.py [hunk 2]
```diff
@@ -18,6 +18,61 @@
...
+
+
+class StaircasingConformalMeshSpec(ConformalMeshSpec):
+    """Simple staircasing scheme based on
+    [Taflove, The Electrical Engineering Handbook 3.629-670 (2005): 15.].
+    """
+
+
+class HeuristicConformalMeshSpec(ConformalMeshSpec):
+    """Slight improvement from the staircasing scheme where the field component near PEC
```
Comment (momchil-flex) [inline tidy3d/components/grid/grid_spec.py:43]: Are you claiming it's an improvement based on sphere results?

One thing that bothers me about the heuristic mesh spec is that the PEC boundary is not effectively continuous. Generally a very unpredictable handling...

But, it's good to have some official description about it as this is the legacy handling. :) I think this description is OK.

**Review by momchil-flex — APPROVED** [review]
Looks good, I guess weather the mode solver changes are good will depend largely on tests? By the way have you thought about whether mode solver PML need to be modified?

File: tidy3d/components/grid/grid_spec.py (hunk 2 above)
Comment (weiliangjin2021) [reply tidy3d/components/grid/grid_spec.py:43]: Now after running more tests, I would remove "slight improvement". In sphere case, it improves, but not in many other cases.

**Comment by weiliangjin2021** [comment]
> By the way have you thought about whether mode solver PML need to be modified?

The split curl efficient is applied directly to permittivity, while in the backend it's still in the curl. So there is no need to modify curl here.

**Review by tylerflex — APPROVED** [review]

**Pull request merged**